nemo/nemo/collections/nlp/modules/common/megatron/bigcode_module.py

```diff
@@ -0,0 +1,3438 @@
+from nemo.collections.nlp.modules.common.megatron.fused_bias_gelu import fused_bias_gelu
+from nemo.collections.nlp.modules.common.megatron.fused_layer_norm import get_layer_norm
+from nemo.collections.nlp.modules.common.megatron.fused_softmax import MatchedScaleMaskSoftmax
+from nemo.collections.nlp.modules.common.megatron.layer_norm_1p import LayerNorm1P
+from nemo.collections.nlp.modules.common.megatron.layer_type import LayerType
+from nemo.collections.nlp.modules.common.megatron.module import MegatronModule
+from nemo.collections.nlp.modules.common.megatron.rotary_pos_embedding import RotaryEmbedding, apply_rotary_pos_emb
+from nemo.collections.nlp.modules.common.megatron.utils import ApexGuardDefaults, attention_mask_func, erf_gelu, get_linear_layer
+from nemo.collections.nlp.modules.common.megatron.utils import openai_gelu as openai_gelu_func
+from nemo.core import adapter_mixins
+from nemo.utils import logging
+from transformers.utils import is_torch_tpu_available
+
+try:
+    from apex.transformer import parallel_state, tensor_parallel
+    from apex.transformer.enums import AttnMaskType, AttnType, ModelType
+    from apex.transformer.utils import divide as safe_divide
+    from apex.transformer.parallel_state import get_tensor_model_parallel_world_size
+    # from apex.normalization import MixedFusedRMSNorm
+    from apex.transformer.layers.layer_norm import FastRMSNorm as MixedFusedRMSNorm
+    from apex.transformer.functional.fused_softmax import FusedScaleMaskSoftmax
+
+    HAVE_APEX = True
+
+except (ImportError, ModuleNotFoundError):
+
+    HAVE_APEX = False
+
+    # fake missing classes with None attributes
+    ModelType = AttnMaskType = AttnType = LayerType = ApexGuardDefaults()
+
+try:
+    from transformer_engine.pytorch import TransformerLayer, fp8_autocast
+    from transformer_engine.common import recipe
+    from transformer_engine.pytorch.distributed import checkpoint as te_checkpoint
+
+    HAVE_TE = True
+
+except:
+    HAVE_TE = False
+
+    # fake missing class
+    class TransformerLayer(ApexGuardDefaults):
+        def __init__(self):
+            super().__init__()
+
+            logging.warning(
+                "Transformer Engine was not found. transformer_engine.pytorch.transformer.TransformerLayer will not work. Please see the NeMo README for installation instructions: https://github.com/NVIDIA/NeMo#megatron-gpt."
+            )
+
+""" We use the following notation throughout this file:
+     h: hidden size
+     n: number of attention heads
+     p: number of model parallel partitions
+     np: n/p
+     hp: h/p
+     hn: h/n
+     b: batch size
+     s: sequence length
+     l: number of layers
+    Transformer takes input of size [s, b, h] and returns a
+    tensor of the same size. We use the following arguments:
+        hyperparameters: transformer hyperparameters
+"""
+
+
+class ParallelMLP(MegatronModule, adapter_mixins.AdapterModuleMixin):
```

Comment:
I dont understand why this entire file is necessary. This appears to be the entire transformer file duplicated.

Reply:
That's a very good remark. My initial goal was to keep it isolated from the GPT-2 implementation. But I put some effort and managed to streamline the code a bit by reusing stuff from the main `module.py` and `transformer.py` that wouldn't affect the BigCode part. Also, I added `BIgCode` to the name of the new classes to highlight the difference. I think that the only other option would be to integrate this directly into the GPT-2 implementation but I think it would add confusion since the way BigCode implemented multi-query attention is not very efficient and I believe GQA is the way to go in the long run.